content/blog/en/dojo-middleware.md

```diff
@@ -0,0 +1,213 @@
+
+Middleware is created with the same factory used to create a widget, except instead of returning a vnode, you return an object or function that defines the API of the middleware and can be used to do some extra work for your widget.
+
+The Dojo documentation touches on [creating your own middleware](https://dojo.io/learn/middleware/middleware-fundamentals#creating-middleware). How could you implement your own custom middleware for your own widgets?
+
+## Validation Middleware
+
+Maybe we're building some form based widgets and we want to provide my own validation. For example, we might want to validate that a phone number is entered correctly.
+
+In this case, we're interested in wrapping an input in some form of validation. Firstly we need to create a `PhoneValidator` widget to wrap DOM `input`. The result would look something like this.
+
+> src/widgets/PhoneNumber.tsx
+
+```tsx
+import { create, tsx } from '@dojo/framework/core/vdom';
+import icache from '@dojo/framework/core/middleware/icache';
+
+import PhoneValidator from './PhoneValidator';
+
+import * as css from './styles/PhoneNumber.m.css';
+
+const factory = create({ icache });
+
+export const PhoneNumber = factory(function PhoneNumber({ middleware: { icache } }) {
+	// use local phone value to pass to validator
+	const phone = icache.getOrSet('phone', '');
+	return (
+		<PhoneValidator phone={phone}>
+			<input
+				placeholder="Enter phone number"
+				pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}"
+				required
+				classes={[css.root]}
+				type="tel"
+				onkeyup={(e) => {
+					icache.set('phone', (e.target as HTMLInputElement).value);
+				}}
+			/>
+		</PhoneValidator>
+	);
+});
+
+export default PhoneNumber;
+```
+
+The idea here is that we want the `PhoneValidator` to place a red or green outline to my input to let me know if it's a valid phone number or not. It's pretty simple, but is something that could be reused across multiple applications.
+
+> src/widgets/PhoneValidator.tsx
+
+```tsx
+import { create, tsx } from '@dojo/framework/core/vdom';
+import phoneNumberMiddleware from '../middleware/phoneNumberMiddleware';
+
+import * as css from './styles/PhoneValidator.m.css';
+
+interface Properties {
+	phone: string;
+}
+
+const factory = create({ phoneNumberMiddleware }).properties<Properties>();
+
+export const PhoneValidator = factory(function PhoneValidator({ children, middleware: { phoneNumberMiddleware } }) {
+	const { valid, value } = phoneNumberMiddleware();
+	let validCss = '';
+	if (value.length) {
+		validCss = valid ? css.valid : css.invalid;
+	}
+	return <div classes={[css.root, validCss]}>{children()}</div>;
+});
+
+export default PhoneValidator;
+```
+
+The `PhoneValidator` uses some middleware that returns the a `valid` property that is either `true` or `false`. It will also return the `value` of the phone number that was tested. Based on whether the phone number is valid or not, it will use some CSS for a red or green border.
```

Comment:
```suggestion
The `PhoneValidator` uses middleware that returns the a `valid` property that is either `true` or `false`. The middle will also return the `value` of the phone number that was tested. Based on whether the phone number is valid or not, it will provide CSS for a red or green border.
```